### PR TITLE
Implement SimpleVocab.

### PR DESCRIPTION
**Pull request by sebpuetz**
Implement the SimpleVocab chunk. The SimpleVocab also acts as the
baseclass for other vocabulary types as their functionality is a
superset of SimpleVocab's methods.

-----------

I had an abstract `Vocab` in my original implementation type that was the base class for all vocabulary types. While rewriting this, I realized that the subword vocabs could just be subclasses of `SimpleVocab` since they all implement at least the functionality of `SimpleVocab` and all that's required would be overriding a few methods.

But I'm not sure if it's confusing that `SimpleVocab` is the common denominator for all vocabularies, do you think this is a sensible solution?

### CODE REVIEW

File: src/finalfusion/vocab/__init__.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    Returns
+    -------
+    vocab : SimpleVocab
+        Returns the first SimpleVocab in the file.
+    """
+    with open(path, "rb") as file:
+        chunk = find_chunk(file, [ChunkIdentifier.SimpleVocab])
+        if chunk is None:
+            raise ValueError('File did not contain a SimpleVocab}')
```
Comment (danieldk) [inline src/finalfusion/vocab/__init__.py]: Stray `}`?

File: src/finalfusion/vocab/simple_vocab.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    @staticmethod
+    def read_chunk(file) -> 'SimpleVocab':
+        length = SimpleVocab._read_binary(file, "<Q")[0]
+        words, index = SimpleVocab._read_items(file, length)
+        return SimpleVocab(words, index)
+
+    def write_chunk(self, file):
+        SimpleVocab._write_binary(file, "<I", int(self.chunk_identifier()))
+        b_words = [bytes(word, "utf-8") for word in self.words]
```
Comment (danieldk) [inline src/finalfusion/vocab/simple_vocab.py]: Perhaps there a better way than storing the full vocabulary in memory? In Rust UTF-8 encoding/decoding is very cheap, how much does it cost in Python? I guess I would mind less if saving took a few milliseconds more than suddenly using whatever amount of memory extra (especially for lower-memory applications).

File: tests/test_vocab.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    assert "1" in v
+    assert None not in v
+    assert ["1", "2"] in v
+    assert {"1", "2"} in v
+    assert map(str, range(2)) in v
+    assert not range(2) in v
+
+
+def test_simple_roundtrip(tests_root, tmp_path):
+    filename = os.path.join(tmp_path, "write_simple.fifu")
```
Comment (danieldk) [inline tests/test_vocab.py]: Things like `tmp_path / "write_simple.fifu"` should work. The same for other pathlib paths. Some people like to overload their operators :).

https://docs.python.org/3.8/library/pathlib.html#operators

**Review by danieldk — CHANGES_REQUESTED** [review]

**Comment by danieldk** [comment]
> But I'm not sure if it's confusing that `SimpleVocab` is the common denominator for all vocabularies, do you think this is a sensible solution?

I am a firm believer of [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance). Using inheritance to reuse code leads to code that is difficult to understand, where you have to step through a class hierarchy or worse, multiple hierarchies, to understand the behavior of a class.

This was one of the important lessons of object-oriented design over the last 30 years or so and why newer languages (i.e. Rust and Go) have made this impossible by removing implementation inheritance altogether, solely relying on interfaces and composition. In traditional OO languages, this means that you only derive from classes that have no implementations and thus act as an interface or at most (like Rust) implementations that are defined in terms of other methods of the interface.

So, in this particular case I'd be in favor of having a `Vocab` that acts as an interface and have the different vocabularies inherit from that directly.

**Comment by sebpuetz** [comment]
Thanks for the feedback!

> 
> 
> > But I'm not sure if it's confusing that `SimpleVocab` is the common denominator for all vocabularies, do you think this is a sensible solution?
> 
> I am a firm believer of [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance). Using inheritance to reuse code leads to code that is difficult to understand, where you have to step through a class hierarchy or worse, multiple hierarchies, to understand the behavior of a class.
> 
> This was one of the important lessons of object-oriented design over the last 30 years or so and why newer languages (i.e. Rust and Go) have made this impossible by removing implementation inheritance altogether, solely relying on interfaces and composition. In traditional OO languages, this means that you only derive from classes that have no implementations and thus act as an interface or at most (like Rust) implementations that are defined in terms of other methods of the interface.
> 
> So, in this particular case I'd be in favor of having a `Vocab` that acts as an interface and have the different vocabularies inherit from that directly.

Makes sense, I should find some time to overhaul the implementation after my lecture tomorrow. I'll update the PR and comment again once it's done!

File: src/finalfusion/vocab/simple_vocab.py (hunk 2 above)
Comment (sebpuetz) [reply src/finalfusion/vocab/simple_vocab.py]: I changed this to encode on the fly and take the sum without storing the words in bytes. I agree, waiting a little longer is probably a nicer scenario than surprising someone with a blown up memory footprint.

Afaik, Python internally uses fixed width representations for characters to allow constant time indexing, so there must be some overhead for conversion, but I doubt that it's enough to actually matter.

**Comment by sebpuetz** [comment]
> Makes sense, I should find some time to overhaul the implementation after my lecture tomorrow. I'll 
> update the PR and comment again once it's done!

Nevermind that, went faster than expected. Should be ready to go now!

File: tests/conftest.py [hunk 4]
```diff
@@ -1,7 +1,17 @@
...
 @pytest.fixture
 def tests_root():
-    yield os.path.dirname(__file__)
+    yield pathlib.PurePath(os.path.dirname(__file__))
+
+
+@pytest.fixture
+def simple_vocab_fifu(tests_root):
+    yield finalfusion.vocab.load_vocab(
+        os.path.join(tests_root, "data/simple_vocab.fifu"))
```
Comment (danieldk) [inline tests/conftest.py]: A forward slash should work here too.

**Review by danieldk — APPROVED** [review]
Looks good, one small comment (I think another os.path.join can probably be replaced).

File: tests/conftest.py (hunk 4 above)
Comment (danieldk) [reply tests/conftest.py]: Or maybe it requires changing the common test code. If so, then it should probably go into a separate PR.

Comment (sebpuetz) [reply tests/conftest.py]: I actually snuck in the common test code change by wrapping `os.path.dirname(__file__)` in `PurePath`, just added the "/" here as well.

**Pull request merged**